### PR TITLE
323 SECIRTS model with waning immunity

### LINKED ISSUE #323: Implement new SECIRTS Model including waning immunity
Implement the Model introduced in https://www.medrxiv.org/content/10.1101/2024.03.01.24303602v1
which is an extension of the SECIRVVS model.

### PR DESCRIPTION
**Pull request by HenrZu**
# Changes and Information

Please **briefly list the changes** (main added features, changed items, or corrected bugs) made:

- Added a new model including waning immunity and temporary immunity states build upon the secirvvs model
- New initialization method

If need be, add additional information and what the reviewer should look out for in particular:
## Merge Request - Guideline Checklist

Please check our [git workflow](https://github.com/SciCompMod/memilio/wiki/git-workflow). Use the **draft** feature if the Pull Request is not yet ready to review.

### Checks by code author

- [x] Every addressed issue is linked (use the "Closes #ISSUE" keyword below)
- [x] New code adheres to [coding guidelines](https://github.com/SciCompMod/memilio/wiki/Coding-guidelines)
- [x] No large data files have been added (files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [x] Tests are added for new functionality and a local test run was successful (with and without OpenMP)
- [x] Appropriate **documentation** for new functionality has been added (Doxygen in the code and Markdown files if necessary)
- [x] Proper attention to licenses, especially no new third-party software with conflicting license has been added
- [ ] (For ABM development) Checked [benchmark results](https://github.com/SciCompMod/memilio/wiki/Agent-Based-Model-Development) and ran and posted a local test above from before and after development to ensure performance is monitored.

### Checks by code reviewer(s)

- [x] Corresponding issue(s) is/are linked and addressed
- [x] Code is clean of development artifacts (no deactivated or commented code lines, no debugging printouts, etc.)
- [x] Appropriate **unit tests** have been added, CI passes, code coverage and performance is acceptable (did not decrease)
- [x] No large data files added in the whole history of commits(files should in sum not exceed 100 KB, avoid PDFs, Word docs, etc.)
- [ ] On merge, add 2-5 lines with the changes (main added features, changed items, or corrected bugs) to the merge-commit-message. This can be taken from the **briefly-list-the-changes** above (best case) or the separate commit messages (worst case).

closes #323

### CODE REVIEW

File: cpp/simulations/CMakeLists.txt [hunk 1]
```diff
@@ -3,10 +3,9 @@ if(MEMILIO_HAS_JSONCPP AND MEMILIO_HAS_HDF5)
     target_link_libraries(2020_npis_wildtype PRIVATE memilio ode_secir Boost::filesystem ${HDF5_C_LIBRARIES})
     target_compile_options(2020_npis_wildtype PRIVATE ${MEMILIO_CXX_FLAGS_ENABLE_WARNING_ERRORS})
 
-    add_executable(2021_vaccination_delta 2021_vaccination_sarscov2_delta_germany.cpp)
-    target_link_libraries(2021_vaccination_delta PRIVATE memilio ode_secirvvs Boost::filesystem ${HDF5_C_LIBRARIES})
-    target_compile_options(2021_vaccination_delta PRIVATE ${MEMILIO_CXX_FLAGS_ENABLE_WARNING_ERRORS})
-
+    # add_executable(2021_vaccination_delta 2021_vaccination_sarscov2_delta_germany.cpp)
+    # target_link_libraries(2021_vaccination_delta PRIVATE memilio ode_secirvvs Boost::filesystem ${HDF5_C_LIBRARIES})
+    # target_compile_options(2021_vaccination_delta PRIVATE ${MEMILIO_CXX_FLAGS_ENABLE_WARNING_ERRORS})
```
[the remark this answers is not in the document]
Comment (HenrZu) [reply cpp/simulations/CMakeLists.txt:8]: As stated in the notes to the PR, Im building the model on the existing secirvvs model (mainly to make the review easier). Once everything is fine from the review side, I will move the model to its own directory.

File: cpp/tests/data/pydata/District/immunity_population.txt [hunk 2]
```diff
@@ -0,0 +1,3 @@
+0.04 0.04 0.075 0.08 0.035 0.01
+0.61 0.61 0.62 0.62 0.58 0.41
+0.35 0.35 0.305 0.3 0.385 0.58
```
Comment (reneSchm) [inline cpp/tests/data/pydata/District/immunity_population.txt:3]: A better design is probably to use a matrix to pass the values to the function that uses them. This still allows using a file like this by reading it into a matrix, then passing it to that function. But the file is not *required* to be there (for tests etc.).

File: cpp/examples/ode_secirts.cpp [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+            double timm_ii = 0.0;
+            for (mio::AgeGroup i = 0; i < nb_groups; i++) {
+                timm_pi += result_interpolated.get_value(t_indx)[model.populations.get_flat_index(
+                    {i, mio::osecirts::InfectionState::TemporaryImmunPartialImmunity})];
+                timm_ii += result_interpolated.get_value(t_indx)[model.populations.get_flat_index(
+                    {i, mio::osecirts::InfectionState::TemporaryImmunImprovedImmunity})];
+            }
+            printf("t=%i, timm_pi=%f, timm_ii=%f\n", int(result_interpolated.get_time(t_indx)), timm_pi, timm_ii);
+        }
+    }
```
[the remark this answers is not in the document]
Comment (HenrZu) [reply cpp/examples/ode_secirts.cpp:150]: In general, I can understand your point. However, we have 3 age groups. So we would have to aggregate the Timeseries before. That shouldnt be straightforward.

File: cpp/models/ode_secirts/parameters_io.cpp [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+        immunity_file.close(); //close the file object.
+    }
+
+    // assert that the sum of the cols is 1 (tolerance 1-10)
+    for (size_t i = 0; i < 6; ++i) {
+        assert(std::abs(ans[0][i] + ans[1][i] + ans[2][i] - 1.0) < 1e-10);
+    }
+
+    return ans;
+}
```
Comment (reneSchm) [inline cpp/models/ode_secirts/parameters_io.cpp:262]: I think the entire immunity_(of_?)population should be replaced by a parameter. If you need to read or write it you can still use the serialization of MatrixXd, and you can use the get_default to store your values.
But it has to change from this. One configuration file with mystery numbers is tolerable, but the same file in three different locations is not acceptable.

File: cpp/models/ode_secirts/parameters_io.h [hunk 5]
```diff
@@ -0,0 +1,1376 @@
...
+IOResult<void> read_confirmed_cases_data(
+    std::string const& path, std::vector<int> const& vregion, Date date, std::vector<std::vector<double>>& num_Exposed,
+    std::vector<std::vector<double>>& num_InfectedNoSymptoms, std::vector<std::vector<double>>& num_InfectedSymptoms,
+    std::vector<std::vector<double>>& num_InfectedSevere, std::vector<std::vector<double>>& num_icu,
+    std::vector<std::vector<double>>& num_death, std::vector<std::vector<double>>& vnum_timm_i,
+    const std::vector<std::vector<int>>& t_Exposed, const std::vector<std::vector<int>>& t_InfectedNoSymptoms,
+    const std::vector<std::vector<int>>& t_InfectedSymptoms, const std::vector<std::vector<int>>& t_InfectedSevere,
+    const std::vector<std::vector<int>>& t_InfectedCritical, const std::vector<std::vector<int>>& vt_imm_interval_i,
+    const std::vector<std::vector<double>>& mu_C_R, const std::vector<std::vector<double>>& mu_I_H,
+    const std::vector<std::vector<double>>& mu_H_U, const std::vector<double>& scaling_factor_inf);
```
Comment (reneSchm) [inline cpp/models/ode_secirts/parameters_io.h:67]: This interface looks honestly terrifying. Some of the parameters look like they are from the model, can you maybe pass osecirts::parameters instead?

Also, this is quite a big function, and you don't explain what it does, or why. Especially the overload of this is a misnomer, it does not read, instead it computes - something?
Similar for e.g. read_divi_data and read_population_data.

**Review by reneSchm — COMMENTED** [review]
I had a quick look at a few files.

File: cpp/models/ode_secirts/parameters_io.h [hunk 6]
```diff
@@ -0,0 +1,1376 @@
...
+    * @param population_data_path path to population data file
+    * @param immunity_data_path path to immunity data file
+    */
+template <class Model>
+IOResult<void> export_input_data_county_timeseries(
+    const std::vector<Model>& model, const std::string& dir, std::vector<int> const& region, Date start_date,
+    const std::vector<double>& scaling_factor_inf, double scaling_factor_icu, int num_days,
+    const std::string& divi_data_path, const std::string& confirmed_cases_path, const std::string& population_data_path,
+    const std::string& immunity_data_path)
+{
```
Comment (reneSchm) [inline cpp/models/ode_secirts/parameters_io.h:698]: This function is way too large. It reads, modifies, and writes data, all in one giant block. You can split it up into at least these three steps, maybe even more. That will allow you to actually test that the modification step works as intended, which you state in TestOdeSECIRTS.export_time_series_init is currently not possible.

Further, after splitting the function up, you don't have to add or modify any files in the test data directory, which I want to reduce as much as possible in the long run. Instead, you can use small test cases (i.e. pairs of input and expected output) to test distinct functionalities of your function. This way, the function is not only covered by a test, but actually tested.

In summary, please split up the function and rewrite the associated test(s). All of this may also apply to some of the other huge functions you use for data handling in this model.

File: cpp/tests/test_odesecirts.cpp [hunk 7]
```diff
@@ -0,0 +1,1654 @@
...
+
+    // Check each flow individually
+    EXPECT_EQ((flows[model.get_flat_flow_index<mio::osecirts::InfectionState::SusceptibleNaive,
+                                               mio::osecirts::InfectionState::ExposedNaive>({mio::AgeGroup(0)})]),
+              0.09375);
+
+    EXPECT_EQ((flows[model.get_flat_flow_index<mio::osecirts::InfectionState::SusceptibleNaive,
+                                               mio::osecirts::InfectionState::TemporaryImmunPartialImmunity>(
+                  {mio::AgeGroup(0)})]),
+              0.90625);
```
Comment (reneSchm) [inline cpp/tests/test_odesecirts.cpp:139]: This test could be shortened using a "compare" vector and a for loop to check all entries. The trade off is a more readable test that makes it easy to verify that all entries are checked, but it would be slightly harder to find out which flow failed a comparison.

File: cpp/memilio/io/epi_data.h [hunk 8]
```diff
@@ -466,16 +468,20 @@ class VaccinationDataEntry
...
-        auto date                       = obj.expect_element("Date", Tag<StringDate>{});
-        auto age_group_str              = obj.expect_element("Age_RKI", Tag<std::string>{});
-        auto state_id                   = obj.expect_optional("ID_State", Tag<regions::StateId>{});
-        auto county_id                  = obj.expect_optional("ID_County", Tag<regions::CountyId>{});
-        auto district_id                = obj.expect_optional("ID_District", Tag<regions::DistrictId>{});
+        auto obj                                   = io.expect_object("VaccinationDataEntry");
+        auto num_vaccinations_partial              = obj.expect_element("Vacc_partially", Tag<double>{});
+        auto num_vaccinations_completed            = obj.expect_element("Vacc_completed", Tag<double>{});
+        auto num_vaccinations_refreshed_first      = obj.expect_element("Vacc_refreshed", Tag<double>{});
+        auto num_vaccinations_refreshed_additional = obj.expect_element("Vacc_refreshed_2", Tag<double>{});
```
[the remark this answers is not in the document]
Comment (HenrZu) [reply cpp/memilio/io/epi_data.h:475]: Depends on the date. For early days, they dont exist. So, i changed them to optional elements.

File: cpp/models/ode_secirts/parameters_io.h (hunk 6 above)
Comment (HenrZu) [reply cpp/models/ode_secirts/parameters_io.h]: i already adjusted the versions of this function for the secir and secirvvs functions.  The former version was indeed horrible :) Now, we are only using tested functions here, so its less prone to errors.

File: cpp/models/ode_secirts/parameters_io.h (hunk 5 above)
Comment (HenrZu) [reply cpp/models/ode_secirts/parameters_io.h]: good point. This was part of the rework of the file. However, we have nearly the same functions multiple times for the bigger ODE models (secir, secirvvs, secirts). I would leave it in this state since this PR has already enough changes. But after this is merged, i will adress this in issue #1108

Comment (reneSchm) [reply cpp/models/ode_secirts/parameters_io.h]: If you add bullet points to that issue for
- changing the interfaces of these read_sth functions,
- adding some more (code) documentation and
- maybe making them more modular,
then I am fine with that.

File: cpp/models/ode_secirts/parameters_io.h [hunk 9]
```diff
@@ -0,0 +1,1158 @@
...
+ * @tparam EpiDataEntry The type of the data entry.
+ * @param data_entry The (RKI) data entry to extract the region ID from.
+ * @return The region ID as integer, or 0 if no specific region information is available.
+ */
+template <class EpiDataEntry>
+int get_region_id(const EpiDataEntry& data_entry)
+{
+    return data_entry.county_id ? data_entry.county_id->get()
+                                : (data_entry.state_id ? data_entry.state_id->get()
+                                                       : (data_entry.district_id ? data_entry.district_id->get() : 0));
```
Comment (reneSchm) [inline cpp/models/ode_secirts/parameters_io.h:61]:
```suggestion
    if (data_entry.county_id) {
        return data_entry.county_id->get();
    }
    if (data_entry.state_id) {
        return data_entry.state_id->get();
    }
    if (data_entry.district_id) {
        return data_entry.district_id->get();
    }
    return 0;
```

File: cpp/models/ode_secirts/parameters_io.h [hunk 10]
```diff
@@ -0,0 +1,1158 @@
...
+* @return An IOResult containing a vector of vectors, where each inner vector represents the population
+*         distribution across age groups for a specific region, or an error if the function fails.
+* @see mio::read_population_data
+* @{
+*/
+IOResult<std::vector<std::vector<double>>> read_population_data(const std::string& path,
+                                                                const std::vector<int>& vregion);
+IOResult<std::vector<std::vector<double>>> read_population_data(const std::vector<PopulationDataEntry>& population_data,
+                                                                const std::vector<int>& vregion);
+/**@}*/
```
Comment (reneSchm) [inline cpp/models/ode_secirts/parameters_io.h:679]: Please split this documentation, as path and population_data are not used by both functions. The rest can be reused.

File: cpp/models/ode_secirts/parameters_io.h [hunk 11]
```diff
@@ -0,0 +1,1158 @@
...
+                                          county, date, scaling_factor_inf, immunity_population));
+    BOOST_OUTCOME_TRY(details::set_population_data(
+        model, path_join(dir, "pydata/Germany", "county_current_population.json"), county, immunity_population));
+
+    if (export_time_series) {
+        // Use only if extrapolated real data is needed for comparison. EXPENSIVE !
+        // Run time equals run time of the previous functions times the num_days !
+        // (This only represents the vectorization of the previous function over all simulation days...)
+        log_warning("Exporting time series of extrapolated real data. This may take some minutes. "
+                    "For simulation runs over the same time period, deactivate it.");
```
Comment (reneSchm) [inline cpp/models/ode_secirts/parameters_io.h:1080]: I would say this should be an info, not warning. But you can add the warning that this is expensive to the param documentation. Same for read_input_data below.
```suggestion
        log_info("Exporting time series of extrapolated real data. This may take some minutes. "
                    "For simulation runs over the same time period, deactivate it.");
```

File: cpp/models/ode_secirts/model.h [hunk 12]
```diff
@@ -0,0 +1,966 @@
...
+        .array() = riskFromInfectedSymptomatic;
+    slice(factors, {Eigen::Index(InfectionState::InfectedSymptomsImprovedImmunity),
+                    Eigen::Index(size_t(params.get_num_groups())), Eigen::Index(InfectionState::Count)})
+        .array() = riskFromInfectedSymptomatic;
+    return factors;
+}
+
+template <typename FP = double, class Base = mio::Simulation<Model<FP>, FP>>
+auto test_commuters(Simulation<FP, Base>& sim, Eigen::Ref<Vector<FP>> migrated, FP time)
+{
```
Comment (reneSchm) [inline cpp/models/ode_secirts/model.h:910]: This function is missing documentation.

**Review by reneSchm — CHANGES_REQUESTED** [review]
Great rework, there are still some fairly large functions, but they are a lot more focused. Most requested changes are about small cleanups.

There are a lot of uncovered lines for error handling, especially in ode_secirts/parameters_io.h. While it would be better to test these functions extensively, I think this is okay, for now. But make a note that this can be improved upon.

File: cpp/memilio/io/epi_data.h [hunk 13]
```diff
@@ -466,16 +468,20 @@ class VaccinationDataEntry
...
     {
-        auto obj                        = io.expect_object("VaccinationDataEntry");
-        auto num_vaccinations_completed = obj.expect_element("Vacc_completed", Tag<double>{});
-        auto date                       = obj.expect_element("Date", Tag<StringDate>{});
-        auto age_group_str              = obj.expect_element("Age_RKI", Tag<std::string>{});
-        auto state_id                   = obj.expect_optional("ID_State", Tag<regions::StateId>{});
-        auto county_id                  = obj.expect_optional("ID_County", Tag<regions::CountyId>{});
-        auto district_id                = obj.expect_optional("ID_District", Tag<regions::DistrictId>{});
+        auto obj                                   = io.expect_object("VaccinationDataEntry");
+        auto num_vaccinations_partial              = obj.expect_element("Vacc_partially", Tag<double>{});
```
[the remark this answers is not in the document]
Comment (HenrZu) [reply cpp/memilio/io/epi_data.h:472]: It was not considered in the vvs model since we calculated them in another way. However, the entry is always part of the vaccination data file.

File: cpp/models/ode_secirts/model.h [hunk 14]
```diff
@@ -0,0 +1,966 @@
...
+                                                     InfectionState::TemporaryImmunImprovedImmunity>({i})] =
+                std::min(y[SIIi] -
+                             flows[this->template get_flat_flow_index<InfectionState::SusceptibleImprovedImmunity,
+                                                                      InfectionState::ExposedImprovedImmunity>({i})],
+                         booster_vaccination[static_cast<size_t>(i)]);
+
+            // ICU capacity shortage is close
+            // TODO: if this is used with vaccination model, it has to be adapted if CriticalPerSevere
+            // is different for different vaccination status. This is not the case here and in addition, ICUCapacity
+            // is set to infinity and this functionality is deactivated, so this is OK for the moment.
```
[the remark this answers is not in the document]
Comment (HenrZu) [reply cpp/models/ode_secirts/model.h:306]: This is surely a problem. I opened Issue #1152. I will address this soon.

**Review by reneSchm — APPROVED** [review]
I think this PR is in a pretty good state now. Unless @mknaranja wants to have a final look, we can merge it soon.

File: cpp/models/ode_secirts/analyze_result.h [hunk 15]
```diff
@@ -0,0 +1,214 @@
+/* 
+* Copyright (C) 2020-2024 MEmilio
+*
+* Authors: Wadim Koslow, Daniel Abele, Martin J. Kühn
```
Comment (reneSchm) [inline cpp/models/ode_secirts/analyze_result.h:4]: @lenaploetzke just spotted that you may need to update the authors for the new files.

Comment (HenrZu) [reply cpp/models/ode_secirts/analyze_result.h]: good point ;) Thank you. Is now updated

File: cpp/models/ode_secirts/README.md [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+| $\lambda_{N,i} =  \rho_{N,i} \sum_j \phi_{i,j}\frac{\xi_{I_{NS}} (I_{NS,N,j} + I_{NS,PI,j} + I_{NS,II,j}) + \xi_{I_{Sy}} (I_{Sy,N,j} + I_{Sy,PI,j}+ I_{Sy,II,j})}{N_j^{D^\perp}}$                      |  `ext_inf_force_dummy`               | Force of infection for susceptibles located in the naive immunity level. |
+| $\lambda_{PI,i} = \rho_{PI,i}\sum_j \phi_{i,j}\frac{\xi_{I_{NS}} (I_{NS,N,j} + I_{NS,PI,j} + I_{NS,II,j}) + \xi_{I_{Sy}} (I_{Sy,N,j} + I_{Sy,PI,j}+ I_{Sy,II,j})}{N_j^{D^\perp}}$                      |  `reducExposedPartialImmunity * ext_inf_force_dummy`               | Force of infection for susceptibles located in the partial immunity level. |
+| $\lambda_{II,i} = \rho_{II}\sum_j \phi_{i,j}\frac{\xi_{I_{NS}} (I_{NS,N,j} + I_{NS,PI,j} + I_{NS,II,j}) + \xi_{I_{Sy}} (I_{Sy,N,j} + I_{Sy,PI,j}+ I_{Sy,II,j})}{N_j^{D^\perp}}$                     |  `reducExposedImprovedImmunity * ext_inf_force_dummy`               | Force of infection for susceptibles located in the improved immunity level. |
+| $\phi$                      |  `ContactPatterns`               | Matrix of daily contact rates / number of daily contacts between different age groups. |
```
Comment (mknaranja) [inline cpp/models/ode_secirts/README.md:18]:
```suggestion
| $\phi$                      |  `ContactPatterns`               | Matrix of daily contact rates, i.e., number of daily contacts between different age groups. |
```

File: cpp/models/ode_secirts/infection_state.h [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+    InfectedSevereImprovedImmunity,
+    InfectedCriticalNaive,
+    InfectedCriticalPartialImmunity,
+    InfectedCriticalImprovedImmunity,
+    SusceptibleImprovedImmunity,
+    DeadNaive,
+    DeadPartialImmunity,
+    DeadImprovedImmunity,
+    TemporaryImmunPartialImmunity,
+    TemporaryImmunImprovedImmunity,
```
Comment (mknaranja) [inline cpp/models/ode_secirts/infection_state.h:66]:
```suggestion
    TemporaryImmunePartialImmunity,
    TemporaryImmuneImprovedImmunity,
```

If this was meant to be short for `Immunity`, OK, but it looks a little bit like wrongly written `Immune` while, then, temporary would have been temporarily.. Don't know 100% how to proceed or rename.

File: cpp/models/ode_secirts/model.h [hunk 18]
```diff
@@ -0,0 +1,971 @@
...
+    Flow<InfectionState::InfectedSymptomsNaive,                       InfectionState::InfectedSevereNaive>,
+    Flow<InfectionState::InfectedSymptomsNaive,                       InfectionState::TemporaryImmunPartialImmunity>,
+    Flow<InfectionState::InfectedSymptomsNaiveConfirmed,              InfectionState::InfectedSevereNaive>,
+    Flow<InfectionState::InfectedSymptomsNaiveConfirmed,              InfectionState::TemporaryImmunPartialImmunity>,
+    Flow<InfectionState::InfectedSevereNaive,                         InfectionState::InfectedCriticalNaive>,
+    Flow<InfectionState::InfectedSevereNaive,                         InfectionState::TemporaryImmunPartialImmunity>, 
+    Flow<InfectionState::InfectedSevereNaive,                         InfectionState::DeadNaive>,
+    Flow<InfectionState::InfectedCriticalNaive,                       InfectionState::DeadNaive>,
+    Flow<InfectionState::InfectedCriticalNaive,                       InfectionState::TemporaryImmunPartialImmunity>,
+    Flow<InfectionState::TemporaryImmunPartialImmunity,                          InfectionState::SusceptiblePartialImmunity>,
```
Comment (mknaranja) [inline cpp/models/ode_secirts/model.h:55]: remove space

File: cpp/models/ode_secirts/model.h [hunk 19]
```diff
@@ -0,0 +1,971 @@
...
+    Flow<InfectionState::InfectedSymptomsImprovedImmunityConfirmed,   InfectionState::TemporaryImmunImprovedImmunity>,
+    Flow<InfectionState::InfectedSevereImprovedImmunity,              InfectionState::InfectedCriticalImprovedImmunity>,
+    Flow<InfectionState::InfectedSevereImprovedImmunity,              InfectionState::TemporaryImmunImprovedImmunity>,
+    Flow<InfectionState::InfectedSevereImprovedImmunity,              InfectionState::DeadImprovedImmunity>,
+    Flow<InfectionState::InfectedCriticalImprovedImmunity,            InfectionState::DeadImprovedImmunity>,
+    Flow<InfectionState::InfectedCriticalImprovedImmunity,            InfectionState::TemporaryImmunImprovedImmunity>,
+    
+    // waning
+    Flow<InfectionState::TemporaryImmunPartialImmunity,                          InfectionState::SusceptiblePartialImmunity>,
+    Flow<InfectionState::TemporaryImmunImprovedImmunity,                          InfectionState::SusceptibleImprovedImmunity>,
```
Comment (mknaranja) [inline cpp/models/ode_secirts/model.h:93]: remove spaces

File: cpp/models/ode_secirts/model.h [hunk 20]
```diff
@@ -0,0 +1,971 @@
...
+        AgeGroup n_agegroups = params.get_num_groups();
+
+        ContactMatrixGroup const& contact_matrix = params.template get<ContactPatterns<FP>>();
+
+        auto icu_occupancy           = 0.0;
+        auto test_and_trace_required = 0.0;
+        for (auto i = AgeGroup(0); i < n_agegroups; ++i) {
+            test_and_trace_required +=
+                (1 - params.template get<RecoveredPerInfectedNoSymptoms<FP>>()[i]) /
+                params.template get<TimeInfectedNoSymptoms<FP>>()[i] *
```
Comment (mknaranja) [inline cpp/models/ode_secirts/model.h:132]: Here is an error. If you want to use the capacity, then you should take into account the reduced time to recover for the more protected states. I think it is also wrong in Wadim's former model. Let's discuss on Monday.

File: cpp/models/ode_secirts/model.h [hunk 21]
```diff
@@ -0,0 +1,971 @@
...
+            auto riskFromInfectedSymptomatic =
+                smoother_cosine(test_and_trace_required, params.template get<TestAndTraceCapacity<FP>>(),
+                                params.template get<TestAndTraceCapacity<FP>>() * 15,
+                                params.template get<RiskOfInfectionFromSymptomatic<FP>>()[i],
+                                params.template get<MaxRiskOfInfectionFromSymptomatic<FP>>()[i]);
+
+            auto riskFromInfectedNoSymptoms =
+                smoother_cosine(test_and_trace_required, params.template get<TestAndTraceCapacity<FP>>(),
+                                params.template get<TestAndTraceCapacity<FP>>() * 2,
+                                params.template get<RelativeTransmissionNoSymptoms<FP>>()[i], 1.0);
```
Comment (mknaranja) [inline cpp/models/ode_secirts/model.h:210]: factors 2 and 15 were made variable in the mean time in the vvs model. Could you adapt? And maybe also check if there were some other adaptations, in particular https://github.com/SciCompMod/memilio/issues/982 and https://github.com/SciCompMod/memilio/pull/1104

**Review by mknaranja — COMMENTED** [review]
I really only did an incomplete review but as @reneSchm did I full one I am OK with merging if we briefly address the open items.

File: cpp/models/ode_secirts/infection_state.h (hunk 17 above)
Comment (HenrZu) [reply cpp/models/ode_secirts/infection_state.h:66]: Initially, `TemporaryImmunPartialImmunity` was used as abbreviation for TemporaryImmunityPartialImmunity. However, i also like the idea to add the 'e' here.

**Review by mknaranja — APPROVED** [review]
